zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1233 @@
+/// - `U`: A getter for the maximum depth of the tree. Using a depth larger than `31` will result in
+///   undefined behavior.
+#[derive(
+    CloneNoBound, Decode, Encode, Eq, MaxEncodedLen, PartialEqNoBound, RuntimeDebugNoBound, TypeInfo,
+)]
+#[scale_info(skip_type_params(T, U))]
+pub struct LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    /// A vector which holds the nodes of the tree. The nodes are ordered by depth (the root is the
+    /// first element of `nodes`) and from left to right. For example, the right-most grandchild of
+    /// the root is at index `6`.
+    pub(crate) nodes: BoundedVec<Node<T>, LiquidityTreeMaxNodes<U>>,
+    /// Maps an account to the node that belongs to it.
+    pub(crate) account_to_index: BoundedBTreeMap<T::AccountId, u32, LiquidityTreeMaxNodes<U>>,
+    /// A vector that contains the indices of abandoned nodes. Sorted in the order in which the
+    /// nodes were abandoned, with the last element of the vector being the most recently abandoned
+    /// node.
+    pub(crate) abandoned_nodes: BoundedVec<u32, LiquidityTreeMaxNodes<U>>,
+}
+
+impl<T, U> LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    /// Create a new liquidity tree.
+    ///
+    /// # Parameters
+    ///
+    /// - `account`: The account to which the tree's root belongs.
+    /// - `stake`: The stake of the tree's root.
+    pub(crate) fn new(
+        account: T::AccountId,
+        stake: BalanceOf<T>,
+    ) -> Result<LiquidityTree<T, U>, DispatchError> {
+        let root = Node::new(account.clone(), stake);
+        let nodes = vec![root]
+            .try_into()
+            .map_err(|_| StorageOverflowError::Nodes.into_dispatch_error::<T>())?;
+        let mut account_to_index: BoundedBTreeMap<_, _, _> = Default::default();
+        account_to_index
+            .try_insert(account, 0u32)
+            .map_err(|_| StorageOverflowError::AccountToIndex.into_dispatch_error::<T>())?;
+        let abandoned_nodes = Default::default();
+        Ok(LiquidityTree { nodes, account_to_index, abandoned_nodes })
+    }
+}
+
+/// Type for nodes of a liquidity tree.
+///
+/// # Notes
+///
+/// - `descendant_stake` does not contain the stake of `self`.
+/// - `lazy_fees`, when propagated, is distributed not only to the descendants of `self`, but also to
+///   `self`.
+#[derive(Clone, Decode, Encode, Eq, MaxEncodedLen, PartialEq, RuntimeDebugNoBound, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub(crate) struct Node<T: Config> {
+    /// The account that the node belongs to. `None` signifies an abandoned node.
+    pub account: Option<T::AccountId>,
+    /// The stake belonging to the owner.
+    pub stake: BalanceOf<T>,
+    /// The fees owed to the owner.
+    pub fees: BalanceOf<T>,
+    /// The sum of the stake of all descendants of this node.
+    pub descendant_stake: BalanceOf<T>,
+    /// The amount of fees to be lazily propagated down the tree.
+    pub lazy_fees: BalanceOf<T>,
```

Comment:
Honestly didn't look at Azuro as close as my adding a reference to it might suggest. Here's why this design requires the distinction:

If you push lazy fees to root (using `deposit_fees`), but root also holds fees, then the two are thrown into the same pot and can't be separated anymore. At least I wouldn't know how.

For example, root has 5 fees and fees have already been propagated. Now 5 lazy fees are pushed and root has 10 in fees. If 20% of the fees belong to root's descendants, this would be 2 units of the total fees stored at root. But actually we have to apply the 20% to the newly arrived lazy fees.

Reply:
I think the Azuro way in this case would be that you compare the stake of the two children of the root to the total stake of the root and then compare how much fees they have and how much fees they should have using the updated fee value. However, this makes me wonder how to ensure that nodes don't get fees that were added before they joined but propagated when they join.
I just realized a test for this is missing, I think we should add it.

Example:
LP0 join 1 stake
trade happens, lazy fees are added
LP1 join 1 stake
LP0 withdraws fees, should get all existing fees.

Reply:
2cc5ce2a09384cb22bba7079862082835ab1d1a7